developer_guides/virtualisation/virtualisation.rst

```diff
@@ -0,0 +1,360 @@
+guest audio interfaces.
+
+3. Implementation
+*****************
+
+3.1. Initialisation
+-------------------
+
+The SOF driver probing on the host remains unmodified, except that additionally
+the vhost driver is also initialised, which registers a newly added
+/dev/vhost-dsp misc device, used by guests to establish VirtQueue links and
+control the SOF vhost driver, using dedicated vhost ioctls.
+
+When QEMU starts a guest instance, it detects the above misc device and
+creates a PCI device for the guest with dedicated vendor and device IDs. All
+vhost PCI devices use the Qumranet / RedHat vendor ID and are allocated a
+device ID from a range, specially donated by RedHat for this purpose.
+
+The SOF VirtIO driver on the guest registers support for that PCI device and
+its .probe() method is called.
+
+During the SOF VirtIO driver probing, the QEMU SOF code reports to the vhost
+driver, which VM image is being instantiated. This is used as a reliable guest
+identification and is needed for SOF guest audio topology selection.
+
+The SOF driver on the guest performs partial initialisation, it skips any
+steps, involving communicating with the actual DSP, e.g. sending the firmware,
+booting it, initialisation the tracing interface.
+
+The VirtIO driver on the guest uses a new SOF_IPC_TPLG_VFE_GET VirtIO-specific
+IPC message to request the topology from the host. The host reads the file,
+corresponding to the guest ID, obtained from the QEMU communication described
+above, from its filesystem, and sends it in multiple chunks to the guest in
+reply to that request.
+
+Guest topology files contain virtual DAI components of type "dai_out" and
+"dai_in." Their data tuples contain a new token of type SOF_TKN_COMP_REF, which
+contains an ID of the connection point, to which this pipeline should be
+attached.
+
+After obtaining the topology, the guest SOF instance uses a second new
+SOF_IPC_TPLG_VFE_COMP_ID VirtIO-specific IPC message to request its allocated
+component ID base. It then uses this base to assign IDs to all of its topology
+components. Those components are then sent to the DSP firmware. This way we
+make sure, that no two components sent to the DSP have the same ID. Note, that
+gaps in component IDs are allowed and don't have any side effects.
+
+The host maintains a list of guest component ID ranges to guarantee, that
+component IDs stay unique regardless of the order in which guests are brought up
+and down.
+
+Next the guest proceeds by sending the parsed topology over IPC to the host,
+which is then forwarded to the DSP firmware. The guest uses the .send_msg()
+callback to forward any IPC, including topology related, with no processing at
+all, to the host. The only exception is blocking IPCs when the guest is
+resuming, while the host doesn't need to be resumed (see section
+`3.3. Deactivation`_ on `reset count`_ for details).
+
+The host receives the guest topology over the IPC virtual queue. In most cases
```

Comment:
the host provides the topology file to the guest
the guest sends topology/configuration IPCs to the host.

which of the two cases does line 255 describe?

Reply:
the latter. I'll try to make it clearer